src/lib.rs

```diff
@@ -218,44 +261,118 @@ impl<T> Sender<T> {
     /// the error involves running any drop implementation on the message type, which might or
     /// might not be lock-free.
     pub fn send(self, message: T) -> Result<(), SendError<T>> {
-        // SAFETY: The channel exists on the heap for the entire duration of this method.
-        let channel: &mut Channel<T> = unsafe { &mut *self.channel_ptr };
+        let channel_ptr = self.channel;
 
         // Don't run our Drop implementation if send was called, any cleanup now happens here
         mem::forget(self);
 
+        // SAFETY: The channel exists on the heap for the entire duration of this method and we
+        // only ever acquire shared references to it. Note that if the receiver disconnects it
+        // does not free the channel.
+        let channel = unsafe { channel_ptr.as_ref() };
+
         // Write the message into the channel on the heap.
-        channel.write_message(message);
+        // SAFETY: The receiver only ever accesses this memory location if we are in the MESSAGE
+        // state, and since we're responsible for setting that state, we can guarantee that we have
+        // exclusive access to this memory location to perform this write.
+        unsafe {
+            channel.write_message(message);
+        }
+
         // Set the state to signal there is a message on the channel.
-        match channel.state.swap(MESSAGE, SeqCst) {
+        // ORDERING: we use release ordering to ensure the write of the message is visible to the
+        // receiving thread. The EMPTY and DISCONNECTED branches do not observe any shared state,
+        // and thus we do not need acquire orderng. The RECEIVING branch manages synchronization
+        // independent of this operation.
+        //
+        // EMPTY + 1 = MESSAGE
+        // RECEIVING + 1 = UNPARKING
+        // DISCONNECTED + 1 = invalid, however this state is never observed
+        match channel.state.fetch_add(1, Release) {
             // The receiver is alive and has not started waiting. Send done.
             EMPTY => Ok(()),
             // The receiver is waiting. Wake it up so it can return the message.
             RECEIVING => {
-                unsafe { channel.take_waker() }.unpark();
+                // ORDERING: Synchronizes with the write of the waker to memory, and prevents the
+                // taking of the waker from being ordered before this operation.
+                fence(Acquire);
```

Comment:
I have now read and understood the UAF issue you described at the start, as well as your proposed fix to it. It's very clever! Thank you for this. I hope and think we can get this integrated somewhat soon.

Reply:
Thanks! It was a pleasure (albeit a head-scratching one) to work on this and get it sorted out :)